js/hview.js

```diff
@@ -632,59 +632,66 @@ $(function () {
         const vListJQueries /* jQuery[] */  = []
         const boxJQueries /* jQuery[] */ = []
         let rootNodes /* ViewNode[] */ = []
+        let nodeMap /* Map<String, ViewNode[]> | null */ = new Map()
 
         viewController.loadViewList().then(function(nodes) {
             rootNodes = nodes
 
-            for(let i = 0; i < rootNodes.length; i++) {
+            function addToNodeMap(node /* ViewNode */, rootNodeIndex /* Integer */) {
+                let mapValue /* ViewNode[] | null */ = nodeMap.get(node.name)
+                if (mapValue == null) {
+                    mapValue = Array(rootNodes.length)
+                    nodeMap.set(node.name, mapValue)
+                }
+                mapValue[rootNodeIndex] = node
+                for (let i = 0; i < node.children.length; i++) {
+                    addToNodeMap(node.children[i], rootNodeIndex)
+                }
+            }
+
+            /* processNextRootNode processes 1 root node at a time. Then it pauses for a few milliseconds
+               before hogging the main thread again so that the user can interact with a working
+               (albeit somewhat slow) UI while the entire collection of time lapse data is being processed.
+
+               If you are wondering why web workers were not used instead, that is because they do not 
+               work well for this use case due to the following: 1) JQuery needs a DOM in order to work
+               properly. I tried using a fake DOM, and JQuery's append/after/appendTo/before methods
+               didn't work. 2) Functions cannot be passed to and from WebWorkers, so the hover / click / etc.
+               methods can't be prepared off the main thread. Also, aside from WebWorkers, Javascript in the
+               browser doesn't provide any alternative methods of multi-threaded programming.*/
+            function processNextRootNode(processedIndex = 0) {
                 const tBox = $("<div>")
                 const tVList = $("<div>")
 
-                renderList(rootNodes[i], tVList, tBox, vListJQueries.length)
+                renderList(rootNodes[processedIndex], tVList, tBox, vListJQueries.length)
                 vListJQueries.push(tVList)
                 boxJQueries.push(tBox)
+                addToNodeMap(rootNodes[processedIndex], processedIndex)
 
                 if ($("#vlist_content label").length == 0) {
                     tVList.children().appendTo("#vlist_content")
                     tBox.children().appendTo("#border-box")
                     applyResizeData();
                 }
+
+                if (processedIndex < rootNodes.length-1) {
+                    setTimeout(processNextRootNode, 16, processedIndex+1);
```

Comment:
instead of using 16 every time, do something like following at the start of the method:

keep processing until the total time is < 10 ms, and then take a break of 6 milliseconds. That would allow faster processing on faster CPUs.
Otherwise this adds an upper limit of the total processing time

Reply:
Unfortunately, it currently takes ~50-80ms to process 1 root node. So "keep processing until the total time is < 10 ms" would be tricky to implement since it would only partially process one root nodes data.

That being said, 16 ms was fairly arbitrary. When I used 1ms instead, I noticed more lag, but not enough that I thought it resulted in a bad user experience, given the processing happening. I will try with 6ms, and see if the lag is any different than 16ms. It might be a sweet spot for processing latency and ux fluidity.

Reply:
Wow 50ms per frame would be difficult to scale. Few things to try:

  renderList sets up some global properties. But since you are just preloading, you should probably use renderNode (with some modification like setting scaleX/scaleY...).  Since the UI will be accessible while the preloading is going on, renderList will keep changing the currentRootNode making the UI flaky.

  Also you can check and store, if the tree in a frame is same as last frame on the worker thread and during rendering you can use node.clone(true) instead of creating the full tree from scratch which should be much faster. You will have to change renderNode to attach the event listeners separately

Reply:
Those are great ideas. The Node.copy will probably work wonders based on my local testing. I don't know about changing renderNode.... I wouldn't expect attaching a listener to take much CPU time. I can double check that though. If changing it speeds things up, I can work on it.

I will also keep thinking about ways to optimize this code.